hpcflow/archive/cloud/dropbox_cp.py

```diff
@@ -0,0 +1,369 @@
+            raise ValueError(f'Specified `local_dir` is not a directory: {dir_to_upload}')
+
+        files_to_upload = generate_files(dir_to_upload, remote_dir, exclude)
+
+        for file in files_to_upload:
+            try:
+                self._archive_file(file)
+            except ArchiveError:
+                print(f"Could not find local file '{file.path.name}' at '{file.path.root}'. "
+                      f"File not archived.")
+                continue
+            except CloudProviderError:
+                print(f"Could not upload file '{file.path.name}' at '{file.path.root}',"
+                      f" due to a Dropbox error. File Not archived.")
+                continue
+
+    def _archive_file(self, file: FileToUpload) -> Union[dropbox.files.Metadata, None]:
+        """Upload a file to a dropbox directory such that if the local file is newer than the
+        copy on Dropbox, the newer file overwrites the older file, and if the local file is
+        older than the copy on Dropbox, the local file is uploaded with an "auto-incremented"
+        name.
+
+        Note that if the modified times are different, the file still won't be uploaded
+        by dropbox if the file contents are identical. The point of checking for the
+        modified times is to save some time by not having Dropbox check file contents.
+
+        Parameters
+        ----------
+        file
+            `FileToUpload` object describing file to be uploaded
+
+        Returns
+        --------
+        FileMetadata of file if file uploaded. If file not uploaded return None.
+        """
+        # Get the last modified time of the file on Dropbox
+        dropbox_modified = self._get_dropbox_file_modified_time(file.dropbox_path)
+        if not dropbox_modified:
+            # File does not exist on dropbox
+            return self._choose_session_type(file)
+        if file.last_modified == dropbox_modified:
+            # If file on Dropbox has the same modified time as the client file then don't upload.
+            return None
+        elif file.last_modified > dropbox_modified:
+            # If the client file is newer than the one on dropbox then overwrite the one on dropbox
+            file.overwrite_mode = dropbox.dropbox.files.WriteMode('overwrite')
+
+        return self._choose_session_type(file)
+
+    def _choose_session_type(self, file: FileToUpload) -> dropbox.files.FileMetadata:
+        """Choose whether to upload the file as a simple upload or a session upload
+        depending on its size.
+
+        Parameters
+        ----------
+        file
+            `FileToUpload` object describing file to be uploaded
+        """
+
+        try:
+            if file.size <= CHUNK_SIZE:
+                return self.simple_upload(file)
+            else:
+                return self.session_upload(file)
+        except Exception as err:
+            raise CloudProviderError(f'Cloud provider error: {err}')
```

Comment:
This is a little confusing, since the method `self._choose_session_type` actually uploads the `file`, which isn't obvious from its name. I would prefer if we remove the `_choose_session_type` method and place its logic inside `_archive_file`:

```suggestion
        dropbox_modified = self._get_dropbox_file_modified_time(file.dropbox_path)
        do_upload = False
        if not dropbox_modified:
            # File does not exist on dropbox
            do_upload = True
        elif file.last_modified > dropbox_modified:
            # If the client file is newer than the one on dropbox then overwrite the one on dropbox
            file.overwrite_mode = dropbox.dropbox.files.WriteMode('overwrite')
            do_upload = True            

        if do_upload:
            try:
                if file.size <= CHUNK_SIZE:
                    return self.simple_upload(file)
                else:
                    return self.session_upload(file)
            except Exception as err:
                raise CloudProviderError(f'Cloud provider error: {err}')        

```

Reply:
For now I have just renamed the function. I think extracting logic to the superclass is easier done when a second provider is added. FileToUpload can move up a level too probably.